fpdf/line_break.py

```diff
@@ -0,0 +1,237 @@
+                len(active_fragment.characters),
+                self.width,
+                self.number_of_spaces,
+            )
+            self.number_of_spaces += 1
+        elif character == SOFT_HYPHEN:
+            self.hyphen_break_hint = hyphen_hint(
+                original_fragment_index,
+                original_character_index,
+                len(self.fragments),
+                len(active_fragment.characters),
+                self.width,
+                self.number_of_spaces,
+                HYPHEN,
+                character_width,
+                style,
+                underline,
+            )
+
+        if character != SOFT_HYPHEN:
+            self.width += character_width
+            active_fragment.characters.append(character)
+
+    def _apply_automatic_hint(self, break_hint):
+        self.fragments = self.fragments[: break_hint.current_line_fragment_index]
+        if self.fragments:
+            self.fragments[-1].trim(break_hint.current_line_character_index)
+        self.number_of_spaces = break_hint.number_of_spaces
+        self.width = break_hint.width
+
+    def manual_break(self, justify=False):
+        return text_line(
+            self.fragments,
+            self.width,
+            self.number_of_spaces,
+            self.number_of_spaces and justify,
+        )
+
+    def automatic_break_possible(self):
+        return self.hyphen_break_hint is not None or self.space_break_hint is not None
+
+    def automatic_break(self, justify):
+        assert self.automatic_break_possible()
+        if self.hyphen_break_hint is not None and (
+            self.space_break_hint is None
+            or self.hyphen_break_hint.width > self.space_break_hint.width
+        ):
+            self._apply_automatic_hint(self.hyphen_break_hint)
+            self.add_character(
+                self.hyphen_break_hint.character_to_append,
+                self.hyphen_break_hint.character_to_append_width,
+                self.hyphen_break_hint.character_to_append_style,
+                self.hyphen_break_hint.character_to_append_underline,
+                self.hyphen_break_hint.original_fragment_index,
+                self.hyphen_break_hint.original_character_index,
+            )
+            return (
+                self.hyphen_break_hint.original_fragment_index,
+                self.hyphen_break_hint.original_character_index,
+                self.manual_break(justify),
+            )
+        self._apply_automatic_hint(self.space_break_hint)
+        return (
+            self.space_break_hint.original_fragment_index,
+            self.space_break_hint.original_character_index,
+            self.manual_break(justify),
+        )
+
+
+class MultiLineBreak:
+    def __init__(self, styled_text_fragments, size_by_style, justify=False):
+
+        self.styled_text_fragments = styled_text_fragments
+
+        self.size_by_style = size_by_style
+        self.justify = justify
+
+        self.fragment_index = 0
```

Comment:
Why are those attributes of a class instance, and not local variables inside `get_line_of_given_width()`?

Reply:
`fragment_index` should definitely be an object attribute, because `get_line_of_given_width` can be called multiple times, and we need to remember how many fragments did this object already parsed.

with `size_by_style` - honestly, I didn't think too much about it. It is technically a function that is used to get the width of the character. I just didn't expect the size of the character to change in the middle of breaking the line to chunks. Is this scenario possible?

`justify` is also unlikely to change during breaking the lines to chunks.

Reply:
I was wondering about `fragment_index` & `character_index`, and with your answer I understand they HAVE to be attributes.
You can leave `self.size_by_style` & `self.justify` as they are, they seem perfectly fine to me.